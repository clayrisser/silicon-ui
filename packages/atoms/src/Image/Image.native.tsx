import React, { FC } from 'react';
import {
  ImageBackground,
  Text as NativeText,
  TouchableOpacity
} from 'react-native';
import {
  background,
  border,
  color,
  layout,
  position,
  shadow,
  space,
  typography
} from 'styled-system';
import {
  ImageProps,
  antiForwardImagePropsKeys,
  splitProps
} from './imageProps';
import { createStyled } from '../styled';

const StyledImageBackground = createStyled<ImageProps>(
  ImageBackground,
  [background, border, color, layout, position, shadow, space, typography],
  antiForwardImagePropsKeys
);

const Image: FC<ImageProps> = (props: ImageProps) => {
  const {
    customImageProps,
    nativeImageProps,
    styledImageProps,
    touchableOpacityProps
  } = splitProps(props);
  const children =
    typeof customImageProps.children === 'string' ? (
      <NativeText>{customImageProps.children}</NativeText>
    ) : (
      customImageProps.children
    );
  const styledImageBackground = (
    <StyledImageBackground
      {...nativeImageProps}
      {...styledImageProps}
      {...customImageProps}
    >
      {children}
    </StyledImageBackground>
  );
  if (Object.keys(touchableOpacityProps).length) {
    return (
      <TouchableOpacity {...touchableOpacityProps}>
        {styledImageBackground}
      </TouchableOpacity>
    );
  }
  return styledImageBackground;
};

Image.defaultProps = {
  height: '100%',
  resizeMode: 'stretch'
};

export default Image;
