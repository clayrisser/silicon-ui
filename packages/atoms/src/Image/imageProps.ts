import { ReactNode, DetailedHTMLProps, ImgHTMLAttributes } from 'react';
import {
  ImageBackgroundProps as ReactNativeImageProps,
  TouchableOpacityProps as ReactNativeTouchableOpacityProps
} from 'react-native';
import {
  BackgroundProps,
  BorderProps,
  ColorProps,
  LayoutProps,
  PositionProps,
  ShadowProps,
  SpaceProps,
  TypographyProps
} from 'styled-system';
import { Theme } from '../themes';

export type DetailedHTMLImageProps = DetailedHTMLProps<
  ImgHTMLAttributes<HTMLImageElement>,
  HTMLImageElement
>;

export interface CustomImageProps {
  children?: ReactNode;
  theme?: Theme;
}

export interface NativeImageProps
  extends Omit<
    ReactNativeImageProps,
    | 'borderBottomLeftRadius'
    | 'borderBottomRightRadius'
    | 'borderRadius'
    | 'borderTopLeftRadius'
    | 'borderTopRightRadius'
    | 'height'
    | 'width'
  > {}

export interface StyledImageProps
  extends BackgroundProps,
    BorderProps,
    ColorProps,
    LayoutProps,
    PositionProps,
    ShadowProps,
    SpaceProps,
    TypographyProps {}

export interface TouchableOpacityProps
  extends Omit<ReactNativeTouchableOpacityProps, 'style'> {}

export interface ImageProps
  extends CustomImageProps,
    NativeImageProps,
    StyledImageProps,
    TouchableOpacityProps {}

export const customImagePropsKeys = new Set<string>(['children', 'theme']);

export const nativeImagePropsKeys = new Set<string>([
  'onPress',
  'onPressIn',
  'onPressOut',
  'source'
]);

export const touchableOpacityPropsKeys = new Set([
  'activeOpacity',
  'delayLongPress',
  'delayPressIn',
  'delayPressOut',
  'disabled',
  'hitSlop',
  'onBlur',
  'onFocus',
  'onLayout',
  'onLongPress',
  'onPress',
  'onPressIn',
  'onPressOut',
  'style',
  'pressRetentionOffset',
  'testID'
]);

// export const antiForwardImagePropsKeys = new Set(['borderColor']);

interface SplitProps {
  customImageProps: CustomImageProps;
  nativeImageProps: NativeImageProps;
  styledImageProps: StyledImageProps;
  touchableOpacityProps: TouchableOpacityProps;
}

export function splitProps(props: ImageProps): SplitProps {
  const customImageProps: { [key: string]: any } = {};
  const nativeImageProps: { [key: string]: any } = {};
  const styledImageProps: { [key: string]: any } = {};
  const touchableOpacityProps: { [key: string]: any } = {};
  Object.entries({ ...props }).forEach(([key, prop]: [string, any]) => {
    if (customImagePropsKeys.has(key)) {
      customImageProps[key] = prop;
    } else if (nativeImagePropsKeys.has(key)) {
      nativeImageProps[key] = prop;
    } else if (touchableOpacityPropsKeys.has(key)) {
      touchableOpacityProps[key] = prop;
    } else {
      styledImageProps[key] = prop;
    }
  });
  return {
    customImageProps,
    nativeImageProps: nativeImageProps as NativeImageProps,
    styledImageProps,
    touchableOpacityProps
  };
}
