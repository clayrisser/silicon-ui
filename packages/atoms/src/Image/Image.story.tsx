import React from 'react';
import { withKnobs, text, number } from '@storybook/addon-knobs';
import { action } from '@storybook/addon-actions';
import Image from './Image';
import Wrapper from '../../storybook/Wrapper';
import austinJpg from './austin.jpg';
import storiesOf from '../../storybook/storiesOf';
import themes from '../themes';
import withThemesProvider from '../../storybook/withThemesProvider';
// import docs from './Image.docs.mdx';

storiesOf('Image', module)
  .addDecorator(withKnobs)
  .addDecorator(withThemesProvider(themes))
  .addParameters({
    // docs: { page: docs },
    // jest: ['Image.spec.tsx']
  })
  .add('with knobs', () => (
    <Wrapper>
      <Image
        source={text('source', austinJpg)}
        alt={text('alt', 'sample image')}
        width={number('width', 400)}
        height={number('height', 250)}
        onPress={action('onPress')}
      />
    </Wrapper>
  ));
