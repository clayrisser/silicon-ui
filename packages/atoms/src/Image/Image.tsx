import React, { FC } from 'react';
import styled, { StyledComponent } from '@emotion/styled';
import {
  background,
  border,
  color,
  compose,
  layout,
  position,
  shadow,
  space,
  typography
} from 'styled-system';
import { ImageProps, DetailedHTMLImageProps, splitProps } from './imageProps';

const HTMLImage: StyledComponent<
  DetailedHTMLImageProps,
  ImageProps,
  object
> = styled.img(
  compose(
    background,
    border,
    color,
    layout,
    position,
    shadow,
    space,
    typography
  )
);

const Image: FC<ImageProps> = (props: ImageProps) => {
  const {
    customImageProps,
    nativeImageProps,
    styledImageProps,
    touchableOpacityProps
  } = splitProps(props);

  function handleClick(e: any) {
    if (props.onPress) props.onPress(e);
  }

  function handleMouseDown(e: any) {
    if (props.onPressIn) props.onPressIn(e);
  }

  function handleMouseUp(e: any) {
    if (props.onPressOut) props.onPressOut(e);
  }

  return (
    <HTMLImage
      {...(customImageProps as any)}
      {...styledImageProps}
      {...touchableOpacityProps}
      onClick={handleClick}
      onMouseDown={handleMouseDown}
      onMouseUp={handleMouseUp}
      src={nativeImageProps.source}
    />
  );
};

Image.defaultProps = {
  source: ''
};

export default Image;
