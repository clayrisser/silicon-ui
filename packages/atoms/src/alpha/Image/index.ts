import Image from './Image';
import { ImageProps } from './imageProps';

export { ImageProps };
export default Image;
