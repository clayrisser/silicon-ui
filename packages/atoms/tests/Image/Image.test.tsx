import React from 'react';
import renderer from 'react-test-renderer';
// import { render, fireEvent } from '@testing-library/react';
import Image from '../../src/Image';

describe('<Image source />', () => {
  it('renders correctly', () => {
    const tree = renderer.create(<Image source="abc" />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
describe('<Image  withProps/>', () => {
  it('renders  with all  props correctly', () => {
    const imageSrc = '';

    const tree = renderer.create(<Image source={imageSrc} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});

// describe('<Image  withProps/>', () => {
//   it('renders  with all  props correctly', () => {
//     const imageSrc = '';

//     const tree = renderer
//       .create(<Image imageSrc={imageSrc} style={{ width: '200px' }} />)
//       .toJSON();
//     expect(tree).toMatchSnapshot();
//   });
// });
