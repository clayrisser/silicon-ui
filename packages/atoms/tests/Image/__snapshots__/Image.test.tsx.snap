// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Image /> renders correctly 1`] = `
<img
  className="css-0"
  onClick={[Function]}
  onMouseDown={[Function]}
  onMouseUp={[Function]}
/>
`;

exports[`<Image source /> renders correctly 1`] = `
<img
  className="css-0"
  onClick={[Function]}
  onMouseDown={[Function]}
  onMouseUp={[Function]}
  src="abc"
/>
`;
